### PR TITLE
Fix recursive views perf

### PR DESCRIPTION
**Pull request by collado-mike**
### Problem
In https://github.com/MarquezProject/marquez/issues/1928, we introduced a `jobs_view`, which used recursive queries to construct a job's fully qualified name by collecting the job's ancestry and joining the names. Additionally, https://github.com/MarquezProject/marquez/issues/1946 introduced job symlinks, which also used recursive queries to find the last link in a chain of symlinks and always return the ultimate target job. Unfortunately, while initial testing showed those queries performed reasonably well on _read_ requests, we didn't sufficiently load test the OpenLineage _write_ API. Unfortunately, under sustained load in a real production environment, the write API (which repeatedly queries the `jobs_view` either directly or via the `runs_view`) puts considerable strain on our Postgres database. 

### Solution
One of the goals with using a view was to consistently construct job FQNs even when ancestors are renamed (using the symlink feature). This allows us to avoid big migration tasks updating the `runs` table with new job names when, e.g., an Airflow DAG or a Spark application is renamed, thus renaming all of its child tasks. However, constructing the FQN on every read is too computationally expensive, so this PR changes the view to compute the FQN on _write_. This continues to use the view, but now there is a `TRIGGER` on `INSERT` to compute the FQN for a job, as well as to traverse the symlink chain and write the new FQN and symlink target to a new table called `jobs_fqn`. This table also collects the past aliases of a job, so we can serve requests that still refer to the old job name, as described in the original [symlink issue](https://github.com/MarquezProject/marquez/issues/1946). 

I utilized writes to the `jobs_view` in order to capture the write by the trigger, so the code now references the view instead of the original `jobs` table. This also allowed me to return the full row from the insert query rather than just the job UUID, as it previously did. I also simplified the uniqueness constraint by adding a `parent_job_id_string` field, which just converts the parent job id to a string or an empty string if null (we couldn't use a null UUID to enforce uniqueness for jobs that have no parents because `NULL` != `NULL` in Postgres).

Given the `jobs_view` and associated write function are likely to change in the future, I made their definitions [repeatable migrations](https://flywaydb.org/documentation/concepts/migrations#repeatable-migrations) so that future changes can be captured in version control (I went ahead and did the same for the `runs_view` for the same reason, though unrelated to the purpose of this PR).

I also included an additional test for the DAG renaming use case.

The following snapshot shows performance in a load test both before and after the proposed code change. Both tests sustained a rate of 900 requests per minute to the OpenLineage write API. I used the same RDS [db.m5.large](https://aws.amazon.com/rds/instance-types/) instance for both tests (cleared the DB structure between tests).

(note that https://github.com/MarquezProject/marquez/pull/2041 is included in the test deployment)

Screen Shot 2022-07-20 at 12 08 43 PM

With the current code at HEAD, the DB's CPU utilization climbs to 95%+ within about an hour, then hovers just below 100%. The request graph shows the request rate dropping after that because pods are consistently being restarted due to failed health checks (can't access the database), so eventually only about 50% of requests are being handled.

Conversely, the DB CPU utilization never rises about ~50% with the new code changes. The request rate stays pretty steady throughout the test (there are still some OOM restarts, but that's consistent with what we see in production with an older version of Marquez). 

> **Note:** All database schema changes require discussion. Please [link the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) for context.

### Checklist

- [X] You've [signed-off](https://github.com/MarquezProject/marquez/blob/main/CONTRIBUTING.md#sign-your-work) your work
- [X] Your changes are accompanied by tests (_if relevant_)
- [X] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [ ] You've updated any relevant documentation (_if relevant_)
- [X] You've updated the [`CHANGELOG.md`](https://github.com/MarquezProject/marquez/blob/main/CHANGELOG.md#unreleased) with details about your change under the "Unreleased" section (_if relevant, depending on the change, this may not be necessary_)
- [X] You've versioned your `.sql` database schema migration according to [Flyway's naming convention](https://flywaydb.org/documentation/concepts/migrations#naming) (_if relevant_)
- [ ] You've included a [header](https://github.com/MarquezProject/marquez/blob/main/CONTRIBUTING.md#copyright--license) in any source code files (_if relevant_)

### CODE REVIEW

File: api/src/main/resources/marquez/db/migration/R__1_jobs_view_and_rewrite_function.sql [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+       f.aliases
+FROM jobs_fqn f,
+     jobs j
+WHERE j.uuid = f.uuid;
+
+
+CREATE OR REPLACE FUNCTION rewrite_jobs_fqn_table() RETURNS TRIGGER AS
+$$
+DECLARE
+    id uuid;
```
Comment (wslulciuc) [inline api/src/main/resources/marquez/db/migration/R__1_jobs_view_and_rewrite_function.sql:29]: _Minor_: For readability, you could us:

```
job_uuid
job_inserted
```

for the function declaration.

File: api/src/main/resources/marquez/db/migration/R__1_jobs_view_and_rewrite_function.sql [hunk 2]
```diff
@@ -0,0 +1,127 @@
+CREATE OR REPLACE VIEW jobs_view
```
Comment (wslulciuc) [inline api/src/main/resources/marquez/db/migration/R__1_jobs_view_and_rewrite_function.sql:1]: _Minor_: I would be more descriptive with the description for the repeatable migration:

> **R__1_add_jobs_view_and_rewrite_jobs_fqn_table_function.sql**

File: api/src/main/resources/marquez/db/migration/R__2_runs_view.sql [hunk 3]
```diff
@@ -0,0 +1,24 @@
+CREATE OR REPLACE VIEW runs_view
```
Comment (wslulciuc) [inline api/src/main/resources/marquez/db/migration/R__2_runs_view.sql:1]: _Minor_: (Similar comment as above) I would be more descriptive with the description and go with:

> **R__1_add_runs_view.sql**

**Review by wslulciuc — APPROVED** [review]
I left a few minors comments, but overall the changes seem reasonable (I double checked the SQL 😉). Thanks for the detailed writeup (as usual), @collado-mike! Also, since the DB _write_ performance issues were observed only after deploying to production, having a not too exhaustive load test in CI with a report can help spot any DB bottlenecks sooner. I've opened https://github.com/MarquezProject/marquez/issues/2047

File: api/src/main/resources/marquez/db/migration/R__1_jobs_view_and_rewrite_function.sql (hunk 2 above)
Comment (collado-mike) [reply api/src/main/resources/marquez/db/migration/R__1_jobs_view_and_rewrite_function.sql:1]: I don't know... after the first time it runs, it's not really _adding_ anything anymore. I kind of like treating the repeatable migration file just like a code file - you'd never use `CreateJobDao.java` as the file name.
Now, I think one legitimate criticism might be whether the `rewrite_function` and the `jobs_view` ought to be in the same migration file (if we're thinking of the file like code). But, IMO, they're so intrinsically connected it made sense to keep them together.

Comment (wslulciuc) [reply api/src/main/resources/marquez/db/migration/R__1_jobs_view_and_rewrite_function.sql:1]: I think that works, I'd like to document how we define migrations (or repeatable ones) for new contributors. So, as long as we have a consistent naming convention, and it's documented then we can leave the name of the file as is (and looks like `R__People_view.sql` is used in Flyways docs on [repeatable migrations](https://flywaydb.org/documentation/tutorials/repeatable#creating-a-repeatable-migration), which is more aligned with your naming approach)

Comment (collado-mike) [reply api/src/main/resources/marquez/db/migration/R__1_jobs_view_and_rewrite_function.sql:1]: Ok, I'll update the files to follow that naming convention and add a line to the `CONTRIBUTING.md` file

**Pull request merged**